server/athenian/api/preloading/cache.py

```diff
@@ -95,6 +97,35 @@ async def refresh(self) -> "CachedDataFrame":
 
         return self
 
+    def filter(
+        self,
+        mask: pd.Series,
+        columns: Optional[Collection[str]],
+        index: Optional[str] = None,
+        uncast: bool = True,
+    ) -> pd.DataFrame:
+        """Filter the wrapped `pandas.Datarame` with the provided mask.
+
+        :param mask: boolean mask for filtering the wrapped DataFrame.
+        :param columns: the columns to select.
+        :param index: the column to set as index.
+        :param uncast: whether to uncast the columns dtypes, see the `_squeeze` method.
+
+        """
+        df = self._df[mask].copy()
```

Comment:
```suggestion
        df = self._df.take(np.nonzero(mask)[0])
```
this works much, much faster.